crates/zkevm_circuits/src/bn254/ec_pairing/implementation.rs

```diff
@@ -0,0 +1,461 @@
+use std::sync::Arc;
+
+use boojum::{
+    gadgets::non_native_field::traits::NonNativeField,
+    pairing::{
+        bn256::{Fq2, FROBENIUS_COEFF_FQ6_C1, XI_TO_Q_MINUS_1_OVER_2},
+        ff::Field,
+    },
+};
+use final_exp::{CompressionMethod, FinalExpEvaluation, HardExpMethod};
+
+use super::*;
+
+// Curve parameter for the BN256 curve
+const SIX_U_PLUS_TWO_WNAF: [i8; 65] = [
+    0, 0, 0, 1, 0, 1, 0, -1, 0, 0, 1, -1, 0, 0, 1, 0, 0, 1, 1, 0, -1, 0, 0, 1, 0, -1, 0, 0, 0, 0,
+    1, 1, 1, 0, 0, -1, 0, 0, 1, 0, 0, 0, 0, 0, -1, 0, 0, 1, 1, 0, 0, -1, 0, 0, 0, 1, 1, 0, -1, 0,
+    0, 1, 0, 1, 1,
+];
+
+/// Struct for the line function evaluation for the BN256 curve (addition and doubling).
+/// The line function is used in the Miller loop of the pairing function.
+pub struct LineFunctionEvaluation<F, CS>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    c0: BN256Fq2NNField<F>,
+    c3: BN256Fq2NNField<F>,
+    c4: BN256Fq2NNField<F>,
+    point: BN256SWProjectivePointTwisted<F>,
+    _marker: std::marker::PhantomData<CS>,
+}
+
+impl<F, CS> LineFunctionEvaluation<F, CS>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    /// Creates a zero instance of the line function evaluation for the BN256 curve.
+    pub fn zero(cs: &mut CS, params: &Arc<BN256BaseNNFieldParams>) -> Self {
+        Self {
+            c0: BN256Fq2NNField::zero(cs, params),
+            c3: BN256Fq2NNField::zero(cs, params),
+            c4: BN256Fq2NNField::zero(cs, params),
+            point: BN256SWProjectivePointTwisted::zero(cs, params),
+            _marker: std::marker::PhantomData::<CS>,
+        }
+    }
+
+    /// Returns the point of the line function evaluation.
+    pub fn point(&self) -> BN256SWProjectivePointTwisted<F> {
+        self.point.clone()
+    }
+
+    /// Returns the coefficients of the line function evaluation.
+    pub fn c0c3c4(&self) -> (BN256Fq2NNField<F>, BN256Fq2NNField<F>, BN256Fq2NNField<F>) {
+        (self.c0.clone(), self.c3.clone(), self.c4.clone())
+    }
+
+    /// This function conducts the doubling step in the Miller loop for the BN256 curve.
+    /// Namely, given `Q` in `E'(Fp2)` and `P` in `E(Fp)`, it computes the line function
+    /// together with the resultant point `T=2*Q`. The implementation is based
+    /// on the _Algorithm 26_ from https://eprint.iacr.org/2010/354.pdf.
+    pub fn doubling_step(
```

Comment:
Remove all this mut ( hint: problem is in mul)

Reply:
I cannot remove mutability without rewriting boojum in core `NonNativeField` trait on which quite a lot of things are dependent